### ISSUE TITLE
Can't make `branch` work

### ISSUE THREAD

**Issue by antoinerey** [acdlite/recompose#109]
I tried to use the `branch` function for 2 hours now, I can't figure out what I'm doing wrong. I'm trying to render a Loader if some props are missing, or the Page component. That's exactly what is done in the docs.

``` javascript

// I create a stateless component as always
const Page = ({ title }) => {
  return (
    <h1>{title}</h1>
  )
}

// I create my Loader component
const Loader = () => {
  return (
    <span>Loading ...</span>
  )
}

// Then, here comes the branch function, to render the relevant one
export default branch(
  props => !! props.title,
  renderComponent(Page),
  renderComponent(Loader)
)

```

I keep getting `ReactCompositeComponent.js:587 Uncaught TypeError: inst.render is not a function`

Am I doing something wrong ? 

What's returned from the `branch` function is a function names `PartialFunc` from `createHelper.js`

PS : I'm using the version **0.15.0**

**Comment by antoinerey** [comment]
I found the solution:

In order to make it work, I had to specify a `BaseComponent` as last argument of the `branch` function. Nevertheless, I still don't get why the exemple from the doc works. (I assume that it works).

Does anyone have an explanation ? :)

**Comment by wuct** [comment]
Because `branch` returns a function which [accepts four arguments](https://github.com/acdlite/recompose/blob/master/src/packages/recompose/branch.js#L5). Without providing `BaseComponent`, what you get is just a function not a React component.

You can use [`lodash/identity`](https://lodash.com/docs#identity), which is a function `arg => arg`, to achieve what you expect in a more semantic way.

```
export default branch(
  props => !! props.title,
  identity,
  renderComponent(Loader)
)(Page)
```

**Comment by antoinerey** [comment]
How does that work in [this exemple](https://github.com/acdlite/recompose/blob/master/docs%2FAPI.md#rendercomponent) then ? The `branch` function is not called with 4 arguments.

**Comment by mrblueblue** [comment]
Just jumping in here: in the end `branch` is called with the four arguments -- `spinnerWhileLoading` seems to be a curried `branch` function. (Please correct if I misunderstood something!)

**Comment by wuct** [comment]
@mrblueblue you are correct.

In the example:

```
const Post = spinnerWhileLoading(

  // Here is the fourth argument
  props => props.title && props.author && props.content,

  // Here is a functional component 
  ({ title, author, content }) => (                                           
    <article>
      <h1>{title}</h1>
      <h2>By {author.name}</h2>
      <div>{content}</div>
    </article>
  )
);
```

**Comment by StevenLangbroek** [comment]
Yeah I'm having issues getting this to work in `compose` flow as well:

```js
// This works, but I have to repeat the component to render after loading:
const preload = (test, Component) => branch(
  test,
  renderComponent(Loader),
  renderComponent(Component)
);

export default compose(
  preload(
    (props) => props.isLoading,
    MyComponent
  ),
  provideHooks(mapHooks),
  reduxForm(mapFormToProps, mapStateToProps, mapDispatchToProps)
)(MyComponent);

// This doesn't work, moved the Component 1 level down so we get it from `compose`. 
// The weird thing is, I do see the Component class (MyComponent in this case) 
// coming down correctly if I log out the args for the returned function...

const preload = (test) => (Component) => branch(
  test,
  renderComponent(Loader),
  renderComponent(Component)
);

export default compose(
  provideHooks(mapHooks),
  reduxForm(mapFormToProps, mapStateToProps, mapDispatchToProps),
  preload(
    (props) => props.isLoading,
  )
)(MyComponent);

/**
Results in this error:
Invariant Violation: Component(...): A valid React element (or null) must be returned. 
You may have returned undefined, an array or some other invalid object.
*/
```

What's going on here? Am I missing something?

**Comment by acdlite** [comment]
@StevenLangbroek You don't need the inner function. This will do:

```js
const preload = (test) => branch(
  test,
  renderComponent(Loader),
  renderComponent(Component)
);
```

Sounds like the confusion arose from two things:

- It may be unclear to some that the 2nd and 3rd arguments to `branch` are higher-order components, not components.
- Currying is confusing. We removed it in 0.17.0 for this reason.

Closing this issue, but if you have any other questions, let us know :)
